### ISSUE TITLE
max-lines suppression should not need to be the first line in the file

### ISSUE THREAD

**Issue by octogonz** [eslint/eslint#12111]
**Tell us about your environment**

* **ESLint Version:** 6.1.0
* **Node Version:** v8.15.0
* **npm Version:** 6.4.1

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

<summary>Configuration</summary>

```js
module.exports = {
    "env": {
        "browser": true,
        "es6": true
    },
    "parserOptions": {
        "ecmaVersion": 2018,
        "sourceType": "module"
    },
    "rules": {
      "max-lines": ["error", { "max": 1000 }]
    }
};
```

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
/*
 * Copyright (c) 2013 ACME, Inc.
 */
/* eslint-disable max-lines */

function f() {
  // (1000+ lines of stuff)
}
```

```bash
eslint ./src/the-file.js
```

**What did you expect to happen?**

The `/* eslint-disable max-lines */` directive should suppress the `max-lines` error.

**What actually happened? Please include the actual, raw output from ESLint.**

The `max-lines` error is reported:  
```
1:1  error  File has too many lines (1247). Maximum allowed is 1000  max-lines
```
If I move `/* eslint-disable max-lines */` to be the first line in the file, then it correctly suppresses the error.  However in our real configuration, we have a rule that wants the copyright header to appear first in the file, so that error is triggered instead.

The only way I could find to satisfy both rules was to do this:
```js
/* eslint-disable */
/*
 * Copyright (c) 2013 ACME, Inc.
 */
/* eslint-enable */

function f() {
  // (1000+ lines of stuff)
}
```

Apparently the `max-lines` and the copyright header rules run when processing the first chunk of text in the file.  So although they technically get reenabled by the `eslint-enable` line, they have missed their chance to execute, and that's why it eliminates the error.  But it means I'm effectively also disabling all other ESLint rules that run in this way.

Seems to me that rules like `max-lines` should run LAST (after all the `eslint-disable` directives have been evaluated) rather than first.  Does that make sense?

**Are you willing to submit a pull request to fix this bug?**

Yes, if someone can explain the right approach

**Comment by platinumazure** [comment]
Hi @octogonz, thanks for the issue.

ESLint disable comments work based on the location of reported lint messages. Lint messages that occur after the disable comment and before any enable comment are filtered out.

The issue in this case is that `max-lines` reports on line 1 column 1, which is properly suppressed in a first-line disable comment but not suppressed if the disable comment is on the second line.

If your goal is to turn off a rule for an entire file, though, you might prefer to use an ESLint rule configuration comment instead. This not only gives you the desired result, but it also prevents the rule from even running on the file, which might save a few cycles.

Please try using this comment instead (anywhere in the file):

```js
/* eslint max-lines: off */
```

I personally find the `/* eslint-disable */` comments to work much better when they are used to avoid linting for a specific section in your file.

**Comment by platinumazure** [comment]
Also: In theory we could change the reporting location in the rule to be the last line, or the n+1 line, which would cover most cases. But I'm not sure I see much value in doing that when it seems that using rule configuration comments works and (in my opinion) makes more sense semantically for what the issue creator wants.

**Comment by octogonz** [comment]
Using `/* eslint max-lines: off */` makes sense and solves this problem.

But I'll mention that from a user experience perspective, the distinction between "disabling" and "configuring" is somewhat subtle.  It's not obvious why `max-lines` can't be disabled like other rules.  Consider that 99% of users aren't the person who initially configured ESLint for their company, so their entire interaction with the tool consists of "disabling" irrelevant rules. A fair amount of rote memorization is already required for that:

- `// eslint-disable-line <rule>`
- `// eslint-disable-next-line <rule>`
- `/* eslint-disable <rule> */`
- Remember NOT to write `// eslint-disable <rule>` because it doesn't work
- Remember `/* eslint-enable */` doesn't undo `/* eslint-disable */`, but instead resets ALL rules

Thus, I wouldn't expect this novice person to have a lot of brain cells left to go read about directives that reconfigure rules.  If they do go down the path, they'll be confronted by a bunch of syntax variations to learn:

- `/* eslint <rule>: "off" */`
- `/* eslint <rule>: off */`
- `/* eslint <rule>: 0 */`
- Weird-seeming arrays like `/* eslint quotes: ["error", "double"] */`
- Also you need to learn that if `/* eslint ... */`  appears multiple times, it's flipping a state for the entire file, unlike `/* eslint-disable <rule> */` which affects a range of lines inside the file

It's a little intimidating at first glance, even though it seems totally obvious to someone who's read the manual and has experience with ESLint.  Maybe the answer is to go RTFM.  But -- from a user experience perspective -- it's good to design software that's simple and intuitive.

Does that make sense?

**Comment by platinumazure** [comment]
> But I'll mention that from a user experience perspective, the distinction between "disabling" and "configuring" is somewhat subtle.

You're absolutely right! I'll note that we're at least exploring whether we can remove the block comment vs line comment distinction [here](https://github.com/eslint/rfcs/pull/34). Otherwise, we would definitely welcome documentation PRs to provide a quick and intuitive introduction to inline comments.

**Comment by nzakas** [comment]
**Note:** This issue appears to have been fixed incidentally by another commit as it's no longer reproducible. We don't know which commit, however.
